qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/data_structures/dns_treemodel.py

```diff
@@ -0,0 +1,394 @@
+            self.setup_model_data(data)
+
+    def index(self, row, column, parent=None):  # overrides QT function
+        if not self.hasIndex(row, column, parent):
+            return QModelIndex()
+        parent_item = self._get_or_create_parent_item(parent)
+        child_item = parent_item.child(row)
+        if child_item:
+            return self.createIndex(row, column, child_item)
+        return QModelIndex()
+
+    def parent(self, index):  # overrides QT function
+        if not index.isValid():
+            return QModelIndex()
+        child_item = index.internalPointer()
+        parent_item = child_item.parent()
+        if parent_item == self.rootItem:
+            return QModelIndex()
+        return self.createIndex(parent_item.row(), 0, parent_item)
+
+    #OKcomment: not used anywhere
+    def headerData(self, section, orientation, role):  # overrides QT function
+        if orientation == Qt.Horizontal and role == Qt.DisplayRole:
+            return self.rootItem.data(section)
+        return None
+
+    def columnCount(self, parent):  # overrides QT function
+        """
+        Returns number of columns.
+        """
+        if parent.isValid():
+            return parent.internalPointer().columnCount()
+        return self.rootItem.columnCount()
+
+    def _scan_command_from_row(self, row):
+        return self.rootItem.child(row).data(0)
+
+    def _scan_expected_points_from_row(self, row):
+        """
+        Returns the number of the scan points which are expected from the scan
+        command can be smaller than number of children if scan did not run
+        completely.
+        """
+        scan_command = self._scan_command_from_row(row)
+        scan_command = scan_command.split('#')[1]
+        if '/' in scan_command:
+            scan_command = scan_command.split('/')[1]
+        if scan_command:
+            return int(scan_command.strip())
+        return 0
+
+    def number_of_scans(self):
+        return self.rootItem.childCount()
+
+    def _scan_range(self):
+        return range(self.number_of_scans())
+
+    @staticmethod
+    def _get_sample_type(sample):
+        if 'vanadium' in sample or 'vana' in sample:
+            return 'vana'
+        if 'nicr' in sample or 'NiCr' in sample:
+            return 'nicr'
+        if 'empty' in sample or 'leer' in sample:
+            return 'empty'
+        return sample
+
+    # complex getting
+    def _get_scan_rows(self):
+        return [
+            row for row in self._scan_range()
+            if 'scan' in self._scan_command_from_row(row)
+        ]
+
+    def _is_scan_complete(self, row):
+        return (self.scan_from_row(row).childCount()
+                >= self._scan_expected_points_from_row(row))
+
+    def text_in_scan(self, row, text):
+        return text in self._scan_command_from_row(row)
+
+    def _get_last_row(self):
+        return self.number_of_scans() - 1
+
+    def data(self, index, role):  # overrides QT function
```

Comment:
could the index be stored as a enum so it is clear what is being called for?

Reply:
I left it unchanged because I don’t see a straightforward way to implement that. Here an abstract tree model for accessing and manipulating data is implemented. In this model, we use a pointer-based tree structure and when we pass a model index to a view, we record the address of the corresponding item in the index and retrieve it later with internalPointer(). One model index may refer to several items, so I am not sure how to implement an enum here.